python/revng/internal/cli/_commands/generate_migrations/__init__.py

```diff
@@ -0,0 +1,239 @@
+
+    def _changes(self, old_schema: Dict, new_schema: Dict) -> List[Tuple]:
+        result: List[GenerateMigrationCommand.Change] = []
+        for name in old_schema["definitions"]:
+            if name in new_schema["definitions"]:
+                old_properties, old_sub_definitions = old_schema["definitions"][name]
+                new_properties, new_sub_definitions = new_schema["definitions"][name]
+
+                for key in self._changed_keys(old_properties, new_properties):
+                    result.append(("definition", "edit", name, key))
+
+                result.extend(
+                    self._single_definition_changes(
+                        name,
+                        old_sub_definitions,
+                        new_sub_definitions,
+                    )
+                )
+            else:
+                result.append(("definition", "remove", name))
+        for name in new_schema["definitions"]:
+            if name not in old_schema["definitions"]:
+                result.append(("definition", "add", name))
+
+        return result
+
+    def _todo_comment(self, change: Change):
+        action: str
+
+        scope, operation, definition_name = change[0], change[1], change[2]
+        if scope == "definition":
+            match operation:
+                case "add":
+                    action = f"addition of {definition_name}"
+                case "remove":
+                    action = f"removal of {definition_name}"
+                case "edit":
+                    # Mypy assumes change has length 3, the minimum possible, so this type: ignore
+                    # is necessary. Same below.
+                    property_name = change[3]  # type: ignore
+                    action = f"change in '{property_name}' of {definition_name}"
+        else:
+            sub_definition_name = change[3]  # type: ignore
+            match operation:
+                case "add":
+                    action = f"addition of {definition_name}.{sub_definition_name}"
+                case "remove":
+                    action = f"removal of {definition_name}.{sub_definition_name}"
+                case "edit":
+                    property_name = change[4]  # type: ignore
+                    action = (
+                        f"change in '{property_name}' of {definition_name}.{sub_definition_name}"
+                    )
+
+        sentence = f"Handle {action}"
+
+        return f"# TODO: {sentence}"
+
+    def _get_jinja_template(self):
+        template_path = Path(__file__).parent / "migration.py.tpl"
+        with open(template_path) as file:
+            template_str = file.read()
+        return jinja2.Template(template_str)
+
+    def _migration_template(self, changes: List[Change]) -> str:
+        template = self._get_jinja_template()
+        return template.render(comments=[self._todo_comment(change) for change in changes])
+
+    def _write_output(self, migration_template: str, output: str | None) -> None:
```

Comment:
There's a fancy way to avoid this:

```python
parser.add_argument('-o', '--output', type=argparse.FileType('w'), default=sys.stdout, help='Output file (default: stdout)')
```

@mrjackv maybe there was some reason not to do this? :thinking:

Reply:
FileType is to be dropped in 3.14, we should stop using it.
I have a wip commit in REP17 which removes it and adds smart functions to substitute its used, but I have yet to push it. I finish it and pass it along to Khaled if needed

Reply:
OK. Then let's just inline this in the caller and drop the error handling.